### PR TITLE
Added Google and MapmyIndia address capture plugins

### PR DESCRIPTION
**Pull request by sachinkum**

### CODE REVIEW

File: dialogy/plugins/text/address_parser/__init__.py [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+                value.value
+                for slot in slots
+                for value in slots[slot].values
+                if slot == "address"
+            ]
+            if not ner_input:
+                return
+
+            most_confident_transcription = ner_input[0]
+            matching_address = get_matching_address_from_gmaps_autocomplete(
```
Comment (ltbringer) [inline dialogy/plugins/text/address_parser/__init__.py:70]: If would benefit to create an AddressEntity like so:

```python
"""
.. _address_entity:
Module provides access to entity types that can be parsed via address using Gmaps or Mapmyindia.

Import classes:
    - AddressEntity
"""
from __future__ import annotations

from typing import Any, Dict, Optional

import attr

from dialogy import constants as const
from dialogy.types.entity.base_entity import BaseEntity
from dialogy.types.entity.deserialize import EntityDeserializer


@EntityDeserializer.register(const.ADDRESS)
@attr.s
class AddressEntity(BaseEntity):
    dim = attr.ib(default="address", kw_only=True)
    entity_type = attr.ib(default="address", kw_only=True)

    @classmethod
    def from_gmaps(
        cls, transcript: str, address_string: str, alternative_index: int, **kwargs: Any
    ) -> AddressEntity:
        value = address_string
        return cls(
            range={const.START: 0, const.END: len(transcript)},
            body=transcript,
            alternative_index=alternative_index,
            values=[{const.VALUE: value}],
        )
```

File: dialogy/plugins/text/address_parser/__init__.py [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+        slots: List[Dict[str, Any]] = first_intent.slots
+
+        if (not intents) or (not isinstance(intents, list)):
+            return
+
+        if not slots:
+            return
+
+        if first_intent.name in self.address_capturing_intents:
+            ner_input = [
```
Comment (ltbringer) [inline dialogy/plugins/text/address_parser/__init__.py:60]: You can add an attribute called `best_transcript` within `Input`. and set its value like so:

```python
    def __attrs_post_init__(self) -> None:
        try:
            object.__setattr__(self, "transcripts", normalize(self.utterances))
            object.__setattr__(self, "best_transcript", fn(self.utterances)) <--- add just this.
        except TypeError:
            ...

```

**Review by ltbringer — CHANGES_REQUESTED** [review]

File: dialogy/plugins/text/address_parser/__init__.py [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+        intents: List[Intent]
+        ner_input: List[str]
+
+        intents = output.intents
+        context = input_
+
+        pincode = None
+
+        if context.history:
+            for intent in context.history[0]["intents"]:
```
[the remark this answers is not in the document]
Comment (sachinkum) [reply dialogy/plugins/text/address_parser/__init__.py]: Needs past context for better predictions

File: dialogy/plugins/text/address_parser/maps.py [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+import googlemaps
+
+from dotenv import load_dotenv
+
+from loguru import logger
+from dialogy.plugins.text.address_parser import mapmyindia
+
+load_dotenv(override=True)
+GOOGLE_MAPS_API_TOKEN = os.getenv("GOOGLE_MAPS_API_TOKEN")
+gmaps = googlemaps.Client(key=GOOGLE_MAPS_API_TOKEN)
```
Comment (ltbringer) [inline dialogy/plugins/text/address_parser/maps.py:15]: Lets expect credentials as input within the plugin for both google and mapmyindia. 

The function `get_matching_address_from_gmaps_autocomplete` can use a fully initialized client. (re-initialization could cause a network overhead).

I would rephrase myself like so:

```python

        provider_functions: Dict[str, Callable[..., str]] = {
            "google": get_matching_address_from_gmaps_autocomplete(gmaps),
            "mmi": get_address_from_mapmyindia_geocode(mmi),
        }
```
and define the provider functions as closures to hold the client.

```python

def get_matching_address_from_gmaps_autocomplete(client: Type):
    def wrapper(
        most_confident_transcription: str,
        country_code: str = "in",
        language: str = "en",
        pin: str = "",
        **kwargs: Any,
    ) -> str:
        # The logic written within `get_matching_address_from_gmaps_autocomplete`
        ...
        return matching_address
   return wrapper
```

**Review by ltbringer — CHANGES_REQUESTED** [review]

File: dialogy/plugins/text/address_parser/maps.py (hunk 4 above)
Comment (ltbringer) [reply dialogy/plugins/text/address_parser/maps.py:15]: I would also avoid `**kwargs` as they are not used.

**Review by ltbringer — APPROVED** [review]
